File: Projects/Pennink-Model/cases/Series5/mf_adapt.m
%% Pennink Series 5 - Marck 1905
% -- FIFTH SERIES OF EXPERIMENTS (MARCH 1905)
%
%see http://www.citg.tudelft.nl/live/pagina.jsp?id=68e12562-a4d2-489a-b82e-deca5dd32c42&lang=en
%
% in this experiment, Pennink studies simultaneous cFresh and saltwater
% flow from a canal in the right and one in the left corner of the sand-box model
% to a canal in the center. The following text is cited from his book, p63:
%
% Pennink writes:
% "To accelerate the motion and to observe the phenomena more quickly and
% more clearly, the slopes are now made steeper. For this purpose a canal
% has been reserved in the middle of hte sand-mass in which a lower level
% may be kept up.
% The discharge from that canal is brought about by means of a pipe
% connected with the back part of the apparatus, the cock of which is
% visible halfway on the left side."
%
% Pennink simulates upconing below the central canal. Interstingly he also
% studied upconing as influenced by a large lateral groundwater flow. This
% gives an interesting shape of the cone. It shows that it is possible for
% the water below the canal to remain fresh, while the salt water enters
% from the side.
%
% You may reduce the computation time by makeing the mesh coarser, say 2 cm
% instead of 1 cm.
%
% TO 090320 100508 100719

clear variables; close all;

basename='PenninkSeries5'; fprintf('Model %s\n',basename);

GREP = 'STRESS PERIOD';

%% Parameters
delta  = 0.03;   % Pennink used milk with density 1.028 - 1.03 kg/L, p48
cFresh = 0;      % Relative conc of cFresh
cSalt  = 1;      % Relative conc of saltwater (milk)
peff   = 0.38;   % [-] calibrated effective porosity
k   =86500/(24*60);       % m/min calibrated conductivity

MW    = 96;      % [cm] width  of model Pennink p63
MH    = 97;      % [cm] height of model Pennink p63
D     = 1.8;     % [cm] thickness of model

% All coordinates in cm relative to LL of sand mass in model

% countor around sand mass
xSand =[
   -3.6406
   97.8254
   97.8254
   46.8008
   -4.2237
];

zSand =[
   -0.7948
   -2.4891
   92.1082
   90.6963
   91.2611
];

% Line along top of sand mass
xTop =[   -5.5143    46.9644    97.7441];
zTop =[   90.6709    90.4615    91.7179];

% countour around right canal
xCanR =[
  100.5756
   86.6065
   90.3819
   92.0809
   94.9125
   96.9890
];

zCanR =[
   97.4417
   97.0229
   87.2506
   82.6437
   80.9685
   80.8987
];
   
% contour around left canal
xCanL =[
   -4.7592
    0.9040
    2.4142
    3.5468
    5.6233
   -4.1929
];

zCanL =[
   81.1081
   81.0383
   81.8061
   86.9016
   96.3947
   96.3947
];

% contour around center (mid) canal
xCanM =[
   41.3012
   43.7552
   44.5103
   45.8317
   48.2858
   49.4184
   50.1735
   51.1174
   51.4949
];

zCanM =[
   96.2551
   80.9685
   76.2219
   74.4769
   74.4769
   75.3145
   77.9670
   88.9957
   97.0229
];

zIFace=23; % initial elevation of interface

milkPoint = [91.5 1 1.5];  % coordates of point where milk was supplied

phiR = 87.5;               % [cm] head in right hand canal
phiS = 87.5-6.5;           % [cm] saline had as mentione by Pennink
Q    = 60e3/60;            % [cm/min, Pennink gives 60 L/h

% Elevation of water table including level in canals
xWT =[
   -3.5555
    3.3253
   44.4849
   50.2397
   90.5236
   98.6555
];

zWT =[
   87.4433
   87.3880
   77.3756
   77.3756
   87.2220
   87.3327
];

%% Model grid
dx = 2.0;              % [cm] grid horizontal cell size
dy =  D;               % [cm] thickness of model as specified by Pennink
dz = 2.0;              % [cm] grid vertical cell size

xGr = 0:dx:MW;         % xGr-grid coordinates
yGr = [0 dy];          % we take the model to be 1 m wide and adjust Pennink's discharge to it
zGr = (MH:-dz:0)';     % zGr line elevations

gr = gridObj(xGr,yGr,zGr);

idxMilk = xyzindex(milkPoint,gr);

ZT = XS( bsxfun(@times, gr.zm(:),interp1(xTop,zTop,gr.xm)));
WT = XS( bsxfun(@times, gr.zm(:),interp1(xWT ,zWT ,gr.xm)));

PhiS = WT(end)-6.5;  % [cm] salwater head according to Pennink

IBOUND = gr.const(0);

iSand = 1;
iMilk = 3;
iCanL = 4;
iCanM = 5;
iCanR = 6;

IBOUND(inpolyz(gr.XM,gr.ZM,xSand,zSand)) = iSand;
IBOUND(inpolyz(gr.XM,gr.ZM,xCanL,zCanL) & gr.ZM<WT) = iCanL;
IBOUND(inpolyz(gr.XM,gr.ZM,xCanM,zCanM) & gr.ZM<WT) = iCanM;
IBOUND(inpolyz(gr.XM,gr.ZM,xCanR,zCanR) & gr.ZM<WT) = iCanR;

IBOUND(idxMilk) = iMilk;

HK = gr.const(k); HK(gr.ZM>WT)=0.1*k;
VK = HK;

PEFF= gr.const(peff); PEFF(PEFF>WT) = peff/3;

STRTHD=WT; STRTHD(gr.ZM<zIFace)=PhiS;

ICBUND = IBOUND;

STCONC = gr.const(cFresh);      % Start with all cells no cSalt
STCONC(gr.ZM<zIFace) = cSalt;   % Make cells below interface salt

%% Head boundary conditions
% Pennink used head boundary conditions in the "canals" of his model. He
% changed these on several moments during the tests. These boundary
% condition are implemented in this simulation in using the CHD package.
% The values are in the accompanying workbook in sheet CHD
% As are the times that the photos were taken.

zoneVals = {iCanL 'hL' 'hL';
            iCanM 'hM' 'hM';
            iCanR 'hR' 'hR';
            iMilk 'hS' 'hS'};
        
zoneConc = {cFresh; cFresh; cFresh; cSalt};
        
[CHD,PNTSRC] = bcnZone(basename,'CHD',IBOUND,zoneVals,zoneConc);

save underneath xSand zSand xCanL zCanL xCanR zCanR xCanM zCanM % need for spi mask

File: Projects/Pennink-Model/cases/Series5/mf_adapt.py
'''Pennink - density flow in Pennnink's (1915) sand box model

FIFTH SERIES OF EXPERIMENTS (MARCH 1905)
See http://www.citg.tudelft.nl/live/pagina.jsp?id=68e12562-a4d2-489a-b82e-deca5dd32c42&lang=en

In this experiment, Pennink (1915) studies simultaneous fresh and saltwater
flow from a canal in the right and one in the left corner of his sand-box model
to a canal in the center.

The following text is cited from his Pennink (1915), p63:

"To accelerate the motion and to observe the phenomena more quickly and
more clearly, the slopes are now made steeper. For this purpose a canal
has been reserved in the middle of the sand mass, in which a lower level
may be kept up.

The discharge from that canal is brought about by means of a pipe
connected with the back part of the apparatus, the cock of which is
visible halfway on the left side."

Pennink simulates upconing below the central canal. Interestingly he also
studied the effect of lateral groundwater flow has on this upconing. Lateral groundwater flow gives an interesting shape of the cone, demonstrating that it is possible for the groundwater below the canal to remain fresh, while the salt water enters from the side.

@TO 090320 100508 100719
@TO 20220116
'''

import os
import numpy as np
from src import mf6tools
from fdm.mfgrid import Grid
import settings

dirs = settings.dirs
sim_name = settings.sim_name
section_name = settings.section_name

# Parameters Excel workbook
params_wbk = settings.params_wbk

# Parameters of the cuuent case model and simulation
pr = settings.props

# Get the data for the packages specific to this simulation.
# Note that the params.wbk sheet NAM determines which packages are used
# and so for which packages data must be provided below.
# Default parameters come from params_wbk.

# %% === tdis ======  time discretization

perDF = mf6tools.get_periodata_from_excel(params_wbk, sheet_name='PER')
period_data = [tuple(sp) for sp in perDF[['PERLEN', 'NSTP', 'TSMULT']].values]

start_date_time = pr['start_date_time'] # Must be string

Simtdis = {'perioddata': period_data,
           'nper': len(period_data),
           'start_date_time': start_date_time,
           'time_units': settings.TIME_UNITS,
           }

# %% === Gwfdis ====== spatial discritization of flow model

x = np.arange(0, pr['L'] + pr['dx'] / 2, +pr['dx'])
z = np.arange(pr['H'], 0 - pr['dz'] / 2, -pr['dz'])
y = [-pr['D'] / 2, pr['D'] / 2]

gr = Grid(x, y, z)

# IDOMAIN: Start all cells inactive vertically flow-through cells
IDOMAIN = gr.const(-1, dtype=int) # Inactive cells are vertially pass-through cells

IDOMAIN[:, 0, :][gr.inpoly(pr[  'sand'], row=0)] = pr['IDSD']
IDOMAIN[:, 0, :][gr.inpoly(pr['canalL'], row=0)] = pr['ICNL'] # Cells in Canal Left
IDOMAIN[:, 0, :][gr.inpoly(pr['canalM'], row=0)] = pr['ICNM'] # Cells in Canal Center
IDOMAIN[:, 0, :][gr.inpoly(pr['canalR'], row=0)] = pr['ICNR'] # Cells in Canal Right

IDOMAIN[gr.ZM < pr['zIface']] = pr['IDMK']

lrcMilk  = gr.lrc_from_xyz(pr['milkInjPnt']) # global coords milk injection point

IDOMAIN.ravel()[gr.Iglob(lrcMilk['ic'])] = pr['iMlkInjPnt'] # mark milk injection cells

Gwfdis = {'gr': gr,
          'idomain': IDOMAIN,
          'length_units': settings.LENGTH_UNITS,
}

# %% === Gwfsto ======= Storage coefficients

Gwfsto = {'sy': gr.const(pr['sy']),
          'ss': gr.const(pr['ss']),
          }

# %% === Gwfnpf ======= Horizontal and vertical conductivity
k = gr.const(pr['k'])

Gwfnpf = { 'k':   k,            
            'icelltype': pr['icelltype'],
}

# %% === Gwfic ======== Initial condictions (head)

hstrt = gr.const(pr['hStrt'])

Gwfic = {'strt': hstrt}

# %% === Gwfchd ======= fixed head

Ix = gr.NOD[0, 0]
IzTopActive = gr.top_active_cells(IDOMAIN)[gr.NOD[0, 0]]

zTop = np.interp(gr.xm, gr.x[[0, int((gr.nx - 1)/2), -1]], pr['sand'][:, 1][2:5])
lrcTopActive = np.vstack((IzTopActive,                 # iz
                          np.zeros(gr.nx, dtype=int),  # iy
                          gr.NOD[0, 0])).T             # ix

# Salt water head of milk injection point
hMilkPnt = pr['zIface'] + (pr['hStrt'] - pr['zIface']) * pr['rhoFresh'] / pr['rhoSalt']

milkOn  = [(tuple(lrc_), hMilkPnt, pr['cSalt']) for lrc_ in lrcMilk['ic']]
milkOff = [(tuple(lrc_), hMilkPnt, pr['cFresh']) for lrc_ in lrcMilk['ic']]

lrcCanL = gr.lrc_from_iglob(gr.NOD[IDOMAIN == pr['ICNL']])
lrcCanM = gr.lrc_from_iglob(gr.NOD[IDOMAIN == pr['ICNM']])
lrcCanR = gr.lrc_from_iglob(gr.NOD[IDOMAIN == pr['ICNR']])

stress_period_data = {0: [],
                      1: (
                        [(lrc, pr['hCNL'], pr['cFresh']) for lrc in lrcCanL] +
                        [(lrc, pr['hCNM'], pr['cFresh']) for lrc in lrcCanM] +
                        [(lrc, pr['hCNR'], pr['cFresh']) for lrc in lrcCanR]),
                      2: (
                        [(lrc, pr['hCNL'], pr['cFresh']) for lrc in lrcCanL] +
                        [(lrc, pr['hCNM'], pr['cFresh']) for lrc in lrcCanM]),
                      3: (
                        [(lrc, pr['hCNM'], pr['cFresh']) for lrc in lrcCanM] +
                        [(lrc, pr['hCNR'], pr['cFresh']) for lrc in lrcCanR]),
                      4:[],
}
  
if False:
   stress_period_data += milkOn
   
Gwfchd ={'auxiliary': 'relconc',
         'stress_period_data': stress_period_data}

# %% === Gwfdrn === drains (can also be used for seepage face).


# %% === Gwfwel === wel (milk injection point)

# %% === Gwfrch === drains (can also be used for seepage face).

# %% === Gwfbuy (boyancy) ====

irhospec = 0
drhdc = (pr['rhoSalt'] - pr['rhoFresh']) / (pr['cSalt'] - pr['cFresh'])
crhoref = pr['cFresh']
modelname = sim_name + 'GWT'
auxspeciesname = "relconc"

Gwfbuy = {'nrhospecies': 1,
          'denseref': pr['rhoFresh'],
          'density_filerecord': os.path.join(dirs.SIM, sim_name + 'Gwf.rho'),
          'packagedata': [irhospec, drhdc, crhoref, modelname, auxspeciesname],
 }


# %% === Gwfoc ==== Output control for flow model
Gwfoc = {'head_filerecord':   os.path.join(dirs.SIM, "{}Gwf.hds".format(sim_name)),
         'budget_filerecord': os.path.join(dirs.SIM, "{}Gwf.cbc".format(sim_name)),
         'saverecord': [("HEAD",   "FREQUENCY", pr['oc_frequency']),
                        ("BUDGET", "FREQUENCY", pr['oc_frequency'])],
}

# %% ============ T R A N S P O R T ====================

#%% === Gwtdis ======= discritization for transport model

# Uses gr above, do not specify

# %% === Gwtfmi ===== Flow model interface

# simultaneous exchange is automatic (see mf_setup.py)

# %% === Gwtmst ===== Mobile storage and transfer

por = gr.const(pr['por'])

Gwtmst = {'porosity': por}

# %% === Gwtadv ====== advection

Gwtadv = {'scheme': 'TVD'} # upstream, central, TVD

# %% === Gwtdsp ====== dispersion & diffusion

Gwtdsp ={**pr['disp']}

# %% === Gwtic ====== initial concentration

cstrt = gr.const(pr['cFresh'])
cstrt[IDOMAIN == pr['IDMK']] = pr['cSalt']

Gwtic = {'strt': cstrt}

# %% === Gwtssm ====== Source-Sink mixing

Gwtssm = {'sources': [['chd', 'AUX', 'relconc'],
                      ['wel', 'AUX', 'relconc']]}

Gwtssm = {'sources': [['chd', 'AUX', 'relconc']]}


# %% === Gwtcnc ====== constant concentraction

# %% === Gwtoc ====== output control

Gwtoc = {
      'concentration_filerecord' : os.path.join(dirs.SIM, '{}Gwt.ucn'.format(sim_name)),
      'budget_filerecord':         os.path.join(dirs.SIM, '{}Gwt.cbc'.format(sim_name)),
      'saverecord' : [("CONCENTRATION", "FREQUENCY", pr['oc_frequency']),
                      ("BUDGET",        "FREQUENCY", pr['oc_frequency'])],
}
print('Done mf_adapt')

if __name__ == '__main__':
   
   print(dirs)     
